File: src/models/resume.rs
use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::fs::read_to_string;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Resume {
	pub name: String,
	pub contacts: Option<ContactList>,
	pub objective: String,
	pub history: TimeLine,
	pub education: TimeLine,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Contact {
	pub href: String,
	pub text: String,
}

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct TimeLineItem {
	pub title: String,
	#[serde(default)]
	pub latest_full_time: bool,
	pub dates: (String, String),
	pub tags: Option<Vec<String>>,
	pub organization: Option<String>,
	pub location: Option<String>,
	pub summary: String,
}

type ContactList = Vec<Contact>;
type TimeLine = Vec<TimeLineItem>;

impl super::Model {
	pub async fn resume(&self) -> Result<Resume> {
		let content = read_to_string(
			self.base_dir.trim_end_matches("/").to_string() + "/content/resume.toml",
		)?;
		let resume = toml::from_str(&content)?;

		Ok(resume)
	}
}

File: models/resume.ts
import * as Toml from "@std/toml";

export type ResumeItem = {
  title: string;
  organization?: string;
  dates: [string, string];
  location?: string;
  tags?: Array<string>;
  summary: string;
};

export type Resume = {
  name: string;
  contacts: Array<
    {
      href: string;
      text: string;
    }
  >;
  objective: string;
  history: Array<ResumeItem>;
  education: Array<ResumeItem>;
};

export async function getResume(): Promise<Resume> {
  const resumeContent = await Deno.readTextFile("./content/resume.toml");
  const resume = Toml.parse(resumeContent) as Resume;

  return resume;
}
